README.md
#dsi2lsl#

Use this program to stream real-time data acquired by a [Wearable Sensing](http://www.wearablesensing.com) DSI headset to the local area network (LAN) via the [LSL](https://github.com/sccn/labstreaminglayer) library.
Below we list the build instructions (including Bluetooth troubleshooting steps) for GNU/Linux, Windows, and Mac. All commands need to be typed on a terminal.

For deployment just copy the subfolder inside Deploy/ that correspond to your OS.

Note: You only need to re-build the app if for some reason the binary in the Deploy folder doesn't run on your system.

#Build#

##GNU/Linux##

1. Move to the dsi2lsl/ folder. Substitute "/.." by the specific full path to the dsi2lsl folder in your file system.

```bash
cd /../dsi2lsl
```

2. Compile the project,

```bash
gcc -DDSI_PLATFORM=-Linux-x86_64  -o "Deploy/Linux/dsi2lsl" dsi2lsl.c  DSI_API_Loader.c -ldl  -L /../dsi2lsl/Deploy/Linux -llsl
```

3. Configure Bluetooth. Pair the DSI device using the Bluetooth device manager. If you don't have the Bluetooth stack properly installed do so using the following command,

```bash
sudo apt-get install bluez bluez-tools
```

4. Find the serial port that corresponds to the device. To accomplish this we need to find the device's address with hcitool,
  and then bind to it,

``` bash
hcitool scan
rfcomm bind /dev/rfcomm0 xx:xx:xx:xx:xx:xx 1
```

where xx:xx:xx:xx:xx:xx represents the output of hcitool (see [this](http://www.westernwillow.com/cms/blog/franco/creating-bluetooth-serial-port-ubuntu) example). We'll need this info later at run time.


##Mac##

1. Move to the dsi2lsl/ folder as before,

```bash
cd /../dsi2lsl
```

2. Compile,

```bash
gcc -DDSI_PLATFORM=-Darwin-x86_64  -o "Deploy/Mac/dsi2lsl" dsi2lsl.c  DSI_API_Loader.c -ldl  -L /../dsi2lsl/Deploy/Mac -llsl64
```

3. Add the DSI device through the Bluetooth manager, then find the port using the following command,

```bash
ls -l tty.*
```
      
for us this returned /dev/tty.DSI7-017-BluetoothSeria. We'll need this info later at run time.


##Windows##

1. Move to the dsi2lsl/ folder as before,

```bash
cd /../dsi2lsl
```

2. Compile the project. We used out of the box MinGW distribution of gcc as the compiler (see a tutorial [here](http://www.mingw.org/wiki/howto_install_the_mingw_gcc_compiler_suite)), although MS Visual C++ and others may work as well,

```bash
gcc -DDSI_PLATFORM=-Windows-x86_32 -o "Deploy/Windows/dsi2lsl.exe" dsi2lsl.c  DSI_API_Loader.c -ldl  -L /../dsi2lsl/Deploy/Windows -llsl32
```

3. Add the device using the Bluetooth manager, then find the port by right clicking on the device and clicking Properties/Hardware. For us the port was COM49. We'll need this info later at run time.


#Examples#

The app runs in the same way on the three platforms. Below we illustrate the different options that can be used. With the exception of the option --help, all the others should be given in --NAME=VALUE format. Use a terminal to run the examples.

Note:
      1. Linux users, since the app uses the serial port to exchange data with the device, your account may need to have superuser privileges (use sudo before any command).
      2. Linux and Mac users may need to add the path the the folder containing the binary and libraries to the system library path, if so use the following command
      
``` bash
export LD_LIBRARY_PATH=/.../dsi2lsl_deploy_folder:$LD_LIBRARY_PATH
```

### --port ###
Specify the serial port with the --port option. If not specified, the API will look for an environment variable called DSISerialPort. This example uses the serial port that the device is binded to on GNU/Linux, for other platforms see the previous secton.

``` bash
./dsi2lsl --port=/dev/rfcomm0
```

### --lsl-stream ###
Specify the name of the LSL outlet that will stream the data to the local network. If omitted, the default name WS-default will be used

``` bash
./dsi2lsl --port=/dev/rfcomm0 --lsl-stream-name=mystream
```

### --montage ###
Specify the montage, a list of channel specifications, comma-separated without spaces, (can also be space-delimited, but then you would need to enclose the option in quotes on the command-line). If omitted all available channels will be used.

``` bash
./dsi2lsl --port=/dev/rfcomm0 --lsl-stream-name=mystream --montage==F3,C3,P3,P4,C4,F4,Pz
```

### --reference ###
Specify the reference, the name of sensor (or linear combination of sensors, without spaces) to be used as reference. Defaults to a \"traditional\" averaged-ears or averaged-mastoids reference if available, or the factory reference (typically Pz) if these sensors are not available.

``` bash
./dsi2lsl --port=/dev/rfcomm0 --lsl-stream-name=mystream --montage==F3,C3,P3,P4,C4,F4,Pz --reference Pz
```


changelog.txt
version:  1.13.1      date:  2016-08-02 15:01:02 -0400

Various internal bug fixes including a fix for a rare "sleep forever" issue.
Internal buffering controller now works robustly.


version:  1.13.0      date:  2015-09-27 20:30:35 -0400

Included capability of collecting the buffered signal
in a background thread, using DSI_Headset_StartBackgroundAcquisition().

Overhauled API documentation/demo program demo.c

Removed dependency on MSVC runtime redistributable under Windows.



version:  1.12.2      date:  2015-07-08 00:21:10 -0400

Minor workaround for firmware quirk.


version:  1.12.1      date:  2015-07-05 18:05:38 -0400

Minor performance and debugging tweaks.


version:  1.12.0      date:  2015-04-14 13:15:27 -0400

Individual one-bit trigger lines are now available as individual sources
TRG0 through TRG7.

New DSI_SourceSelection constant TriggerBits allows selection of the new
sources (they are de-selected by default).

New functions DSI_Channel_IsTrigger( c ) and DSI_Source_IsTrigger( s ) test
whether a channel (or source) is any one of TRG0, .... TRG7, or the old TRG.



version:  1.11.0      date:  2015-02-05 12:39:05 -0500

New function DSI_Headset_GetSensorCount( h ) returns the number of sensors
declared in headset's calibration memory (7 for DSI7s, 24 for DSI24s, if
correctly flashed).

"Feature"-fix:  DSI7 trigger values were previously shifted by 4 bits
relative to DSI24 trigger values from the same input.  This is now equalized.



version:  1.10.0      date:  2015-01-26 23:30:00 -0500

New function DSI_GetDefaultPort() returns the value of the DSISerialPort
environment variable.   DSI_Headset_Connect() updated so that is uses this
default setting automatically if you supply either an empty string or the
string "default" as the port address.



version:  1.9.1       date:  2015-01-23 18:27:42 -0500

Internal fix for rare bugs (changed to handling of string warning/error messages).


version:  1.9.0       date:  2015-01-22 19:03:34 -0500

One entirely internal but significant change: the trigger signal (TRG) is now
delayed by 53.33 msec (16 samples at 300 Hz) relative to previous versions,
to bring it into alignment with the signal. The signal is delayed by this amount
due to the headset's built-in anti-aliasing filter.



version:  1.8.0       date:  2015-01-16 20:56:27 -0500

Added DSI_Headset_SetVerbosity method and improved debugging output under
certain conditions.



version:  1.7.1       date:  2014-12-03 21:00:22 -0500

Changed DSI_*_GetChannelByName functions so that Channels can be matched by the
aliases of their principal Source, as well as by its canonical name.


version:  1.7.0       date:  2014-11-28 04:24:28 -0500

Minor tweaks and additions:

- Added new functions DSI_Headset_GetChannelByName(h, name) and
  DSI_ProcessingStage_GetChannelByName(p, name) for completeness. 
- Channel specifications may now optionally be comma- instead of space-delimited. 
- Better demo code with better documentation (now called demo.c instead of test.c).



version:  1.6.2       date:  2014-10-31 18:32:28 -0400

Headsets are now initialized with the impedance driver turned off by default, so
API client code must now call DSI_Headset_StartImpedanceDriver(h) explicitly if
impedance measurements are needed. Any "locks" on analog-reset mode are also
released during initialization.


version:  1.6.1       date:  2014-10-31 15:16:46 -0400

Fixed Halloween Bug which would cause trouble when using headsets on Oct 31st
(and various other dates throughout the year).


version:  1.6.0       date:  2014-10-26 23:38:52 -0400

Added DSI_ProcessingStage methods as a hook for custom processing.

Renamed DSI_Channel_GetSignal(h) to DSI_Channel_CalculateRawSignal(h). This
makes it more explicit that the signals this returns are calculated from scratch
as a weighted combination of raw DSI_Source signals---this was counterintuitive
when the DSI_Channel in question was not a raw channel but rather a member of a
custom DSI_ProcessingStage.

For backward compatibility, the old name DSI_Channel_GetSignal(h) remains, but
now it delivers the buffered signal by calling DSI_Channel_LookBack(h,0), not
the from-scratch signal. In most cases, existing API client code that calls
DSI_Channel_GetSignal(h) will not be impacted by the difference.



version:  1.5.0       date:  2014-10-17 18:56:32 -0400

Headsets now call DSI_Headset_SetTraditionalReference(h, true) automatically
when they are initialized. This means that the EEG is referenced against
averaged ears or averaged mastoids by default,  rather than against the factory
reference Pz. Previously, API client code had to call
DSI_Headset_SetTraditionalReference explicitly---this is no longer necessary.  A
warning message (level 1) is issued if ears/mastoids are not found during
initialization (i.e. if the headset calibration memory does not contain a record
of ear or mastoid sensors).

"IsEEG" methods have been renamed to "IsReferentialEEG"---i.e. now use
DSI_Source_IsReferentalEEG(s) and DSI_Channel_IsReferentialEEG(c).


dsi2lsl.c
/*

# This file is part of the Application Programmer's Interface (API) for Dry
# Sensor Interface (DSI) EEG systems by Wearable Sensing. The API consists of
# code, headers, dynamic libraries and documentation.  The API allows software
# developers to interface directly with DSI systems to control and to acquire
# data from them.
#
# The API is not certified to any specific standard. It is not intended for
# clinical use. The API, and software that makes use of it, should not be used
# for diagnostic or other clinical purposes.  The API is intended for research
# use and is provided on an "AS IS" basis.  WEARABLE SENSING, INCLUDING ITS
# SUBSIDIARIES, DISCLAIMS ANY AND ALL WARRANTIES EXPRESSED, STATUTORY OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO ANY IMPLIED WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, NON-INFRINGEMENT OR THIRD
# PARTY RIGHTS.
#
# Copyright (c) 2014-2016 Wearable Sensing LLC

This file implements the integration between Wearable Sensing DSI C/C++ API and
the LSL library. This example has been contributed by Qusp based on the demo.c
example brovided by Wearable Sensing.

Copyright (C) 2014-2016 Syntrogi Inc (dba Qusp). All rights reserved.

1. Getting started

To create a DSI application:

(1) Add the file DSI_API_Loader.c to your project (can be compiled as C or C++)

(2) #include "DSI.h" in your own C or C++ code

(3) Call Load_DSI_API() in your code before calling any of the other functions.
    The argument to Load_DSI_API may be a string that explicitly specifies the
    name of the dynamic library file to load. Alternatively, if you pass NULL
    then the default name will be used.  This default can be manipulated at
    compile time by defining the DSI_PLATFORM symbol.

    For example, using Visual C++ to compile a 32-bit binary on a Windows
    machine, you could say:

        cl /DDSI_PLATFORM=-Windows-i386  demo.c DSI_API_Loader.c

    which would cause Load_DSI_API(NULL) to look for "libDSI-Windows-i386.dll"
    instead of just "libDSI.dll".

    As an another example, using gcc on Mac OSX on a 64-bit intel machine:

        gcc -DDSI_PLATFORM=-Darwin-x86_64  demo.c DSI_API_Loader.c

    will cause Load_DSI_API(NULL) to look for "libDSI-Darwin-x86_64.dylib"
    instead of just "libDSI.dylib".

    Finally, on 64-bit Linux:

        gcc -DDSI_PLATFORM=-Linux-x86_64  demo.c DSI_API_Loader.c -ldl -lm

    will cause Load_DSI_API(NULL) to look for "libDSI-Linux-x86_64.so"
    instead of just "libDSI.so"  (note also that Linux compilation will likely
    require you to add linker flags explicitly at the end of the command: the
    -ldl flag is needed in all cases, and -lm is needed for the current demo.c
    because of its reference to the sin function). See also section 7.3 for more
    Linux notes.

(4) Check the return value of Load_DSI_API before proceeding. It will be zero on
    success. Any non-zero value means the API failed to load and other API calls
    (function names starting with DSI_) should not be called.

    If the return value is negative, it means that the libDSI file could not be
    located or opened. If this is the case, check the following rules for locating
    dynamic libraries at runtime:

    * On Windows, the .dll file needs to be either in the current working directory,
      or in one of the directories listed (separated by semicolons) in the Path
      environment variable.

    * On Mac OSX, the .dylib file needs to be either in the current working
      directory, or in one of the directories listed (separated by colons) in the
      DYLD_LIBRARY_PATH or LD_LIBRARY_PATH environment variables.

    * On Linux, the .so file must be in one of the directories listed (separated
      by colons) in the LD_LIBRARY_PATH environment variable (the current working
      directory will not be searched by default).  Currently, your system's version
      of GLIBC (the GNU C Library shared object) must also be 2.14 or later,
      otherwise the libDSI shared library will fail to load. See also the caveats
      and notes in section 7.3 below.

This file may be compiled. In it, we will build documented example applications
for recording signal or impedance values to a .csv file.  The `main()` function,
below, illustrates how to load the API and check for errors.

The file provides demos of various different aspects of API functionality.
The first command-line argument allows you to select which demo mode you are
running The default demo mode is "signals" but the following are recognized:

	version, info, buffering, montages, naming, impedances, signals

Call with the --help flag for global help, or with the help flag following
one of the modes above for mode-specific help.
                                                                              */


                                                                              /*
First let's get the header includes and advance declarations out of the way:
                                                                              */
    ///////////////////////////////////////////////////////////////////////////////

#   include "DSI.h"    // Wearable Sensing DSI API header
#   include "lsl_c.h"  // LSL header

#   include <stdio.h>   // C standard library
#   include <string.h>  // Needed for processing strings
#   include <stdlib.h>  // Needed for allocating memory on demand
#   include <signal.h>  // Needed for implementing the Ctrl+C signaling for exiting the program

///////////////////////////////////////////////////////////////////////////////
// Helper functions and macros that will be defined down bellow
int               StartUp( int argc, const char * argv[], DSI_Headset *headsetOut, int * helpOut );
int                Finish( DSI_Headset h );
int            GlobalHelp( int argc, const char * argv[] );
lsl_outlet        InitLSL( DSI_Headset h, const char * streamName);
void             OnSample( DSI_Headset h, double packetOffsetTime, void * userData);
void      getRandomString( char *s, const int len);
const char * GetStringOpt( int argc, const char * argv[], const char * keyword1, const char * keyword2 );
int         GetIntegerOpt( int argc, const char * argv[], const char * keyword1, const char * keyword2, int defaultValue );

float *sample;
static volatile int KeepRunning = 1;
void  QuitHandler(int a){ KeepRunning = 0;}

#   define REPORT( fmt, x )  fprintf( stderr, #x " = " fmt "\n", ( x ) )
int CheckError( void ){
  if( DSI_Error() ) return fprintf( stderr, "%s\n", DSI_ClearError() );
  else return 0;
}
#   define CHECK   if( CheckError() != 0 ) return -1;
///////////////////////////////////////////////////////////////////////////////

                                                                              /*
Now, our main() function will illustrate how to load the DSI API, assuming
DSI_API_Loader.c has also been compiled and linked into your project:
                                                                              */
int main( int argc, const char * argv[] )
{
  // First load the libDSI dynamic library
  const char * dllname = NULL;
  // dllname = NULL causes the default name for the dynamic library to be used
  // (incorporating the DSI_PLATFORM macro, if defined).

  int load_error = Load_DSI_API( dllname );
  // The macro that Load_DSI_API uses to detect whether the name is NULL, and
  // to substitute the default name if so, is called DSI_DYLIB_NAME. We'll
  // make use of it directly in the error messages:
  if( load_error < 0 ) return fprintf( stderr, "failed to load dynamic library \"%s\"\n", DSI_DYLIB_NAME( dllname ) );
  if( load_error > 0 ) return fprintf( stderr, "failed to import %d functions from dynamic library \"%s\"\n", load_error, DSI_DYLIB_NAME( dllname ) );
  // Print version information - warn if the header and dynamic-library versions don't match
  fprintf( stderr, "DSI API version %s loaded\n", DSI_GetAPIVersion() );
	if( strcmp( DSI_GetAPIVersion(), DSI_API_VERSION ) != 0 ) fprintf( stderr, "WARNING - mismatched versioning: program was compiled with DSI.h version %s but just loaded shared library version %s. You should ensure that you are using matching versions of the API files - contact Wearable Sensing if you are missing a file.\n", DSI_API_VERSION, DSI_GetAPIVersion() );

  // Implements a Ctrl+C signal handler to quit the program (some terminals actually use Ctrl+Shift+C instead)
  signal(SIGINT, QuitHandler);

  printf("Initializing the headset\n");
  DSI_Headset h;
  int help, error;
  error = StartUp( argc, argv, &h, &help ); // common routine for all demos, to initialize the headset
  if( error || help){
    GlobalHelp(argc, argv);
    return error;
  }

  // Initialize LSL outlet
  const char * streamName = GetStringOpt(  argc, argv, "lsl-stream-name",   "m" );
  if (!streamName) streamName = "WS-default";
  printf("Initializing %s outlet\n", streamName);
  lsl_outlet outlet = InitLSL(h, streamName); CHECK; /* stream outlet */

  // Set the sample callback (forward every data sample received to LSL)
  DSI_Headset_SetSampleCallback( h, OnSample, outlet ); CHECK
  // This registers our `DSI_SampleCallback`, to be called on every
  // new sample during `DSI_Headset_Idle()`.

  // Start data acquisition
  printf("Starting data acquisition\n");
  DSI_Headset_StartDataAcquisition( h ); CHECK

  // Start streaming
  printf("Streaming...\n");
  while( KeepRunning==1 ){
    DSI_Headset_Idle( h, 0.0 ); CHECK
  }

  // Gracefully exit the program
  printf("\n%s will exit now.\nGood bye!\n", argv[ 0 ]);
  lsl_destroy_outlet(outlet);
  return Finish( h );
}



void OnSample( DSI_Headset h, double packetOffsetTime, void * outlet)
{
  unsigned int channelIndex;
  unsigned int numberOfChannels = DSI_Headset_GetNumberOfChannels( h );
  if (sample==NULL) sample = (float *)malloc( numberOfChannels * sizeof(float));
  for(channelIndex=0; channelIndex < numberOfChannels; channelIndex++){
    sample[channelIndex] = (float) DSI_Channel_GetSignal( DSI_Headset_GetChannelByIndex( h, channelIndex ) );
  }
	lsl_push_sample_f(outlet, sample);
}

int Message( const char * msg, int debugLevel ){
  return fprintf( stderr, "DSI Message (level %d): %s\n", debugLevel, msg );
}
/*
4. Connecting/disconnecting the headset

The `StartUp()` function below gives examples of `DSI_Headset_...()` function
calls that might commonly be used to initialize the headset.  We'll call this
function in all our subsequent demos.
                                                                              */
int StartUp( int argc, const char * argv[], DSI_Headset * headsetOut, int * helpOut )
{
  DSI_Headset h;

  // First, let's figure out some of the command-line options that affect
  // headset initialization, for the purposes of this demo program.
  // (NB: `GetStringOpt` and `GetIntegerOpt` are not part of the DSI API -
  // they're defined below as part of this demo program.  But you will
  // presumably have your own mechanisms for allowing the user to
  // configure your program.)
  //
  //   "[I]t is more powerful than the sum total of all the computers on
  //    this planet... But over ninety-five percent of that power is used
  //    in simply understanding what it is you want it to do."
  //            --Douglas Adams,  Dirk Gently's Holistic Detective Agency
  int          help       = GetStringOpt(  argc, argv, "help",      "h" ) != NULL;
  const char * serialPort = GetStringOpt(  argc, argv, "port",      "p" );
  const char * montage    = GetStringOpt(  argc, argv, "montage",   "m" );
  const char * reference  = GetStringOpt(  argc, argv, "reference", "r" );
  int          verbosity  = GetIntegerOpt( argc, argv, "verbosity", "v", 2 );
  if( headsetOut ) *headsetOut = NULL;
  if( helpOut ) *helpOut = help;
  if( help ) return 0;

  // OK, now we can begin:
  h = DSI_Headset_New( NULL ); CHECK
  // Passing NULL defers setup of the serial port connection until later...

  DSI_Headset_SetMessageCallback( h, Message ); CHECK
  DSI_Headset_SetVerbosity( h, verbosity ); CHECK
  // ...which allows us to configure the way we handle any debugging messages
  // that occur during connection (see our definition of the `DSI_MessageCallback`
  // function `Message()` above).

  DSI_Headset_Connect( h, serialPort ); CHECK
  // Now we establish the serial port connection and initialize the headset.
  // In this demo program, the string supplied in the --port command-line
  // option is used as the serial port address (if this string is empty, the
  // API will automatically look for an environment variable called
  // DSISerialPort).

  DSI_Headset_ChooseChannels( h, montage, reference, 1 ); CHECK
  // Sets up the montage according to strings supplied in the --montage and
  // --reference command-line options, if any.

  fprintf( stderr, "%s\n", DSI_Headset_GetInfoString( h ) ); CHECK
  // prints an overview of what is known about the headset

  if( headsetOut ) *headsetOut = h;
  if( helpOut ) *helpOut = help;
  return 0;
}
                                                                              /*
While we're at it let's write the corresponding wrap-up steps that allow
the headset to back out gracefully from whatever task it was performing
before finally de-initializing the DSI_Headset object:
                                                                              */
int Finish( DSI_Headset h )
{
  DSI_Headset_SetSampleCallback( h, NULL, NULL ); CHECK
  // This stops our application from responding to received samples.

  DSI_Headset_StopDataAcquisition( h ); CHECK
  // This send a command to the headset to tell it to stop sending samples.

  DSI_Headset_Idle( h, 1.0 ); CHECK
  // This allows more than enough time to receive any samples that were
  // sent before the stop command is carried out, along with the alarm
  // signal that the headset sends out when it stops

  DSI_Headset_Delete( h ); CHECK
  // This is the only really necessary step. Disconnects from the serial
  // port, frees memory, etc.

  return 0;
}


void getRandomString(char *s, const int len)
{
  int i = 0;
  static const char alphanum[] =     "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
  for (i=0; i < len; ++i){ s[i] = alphanum[rand() % (sizeof(alphanum) - 1)];}
  s[len] = 0;
}

lsl_outlet InitLSL(DSI_Headset h, const char * streamName)
{
  unsigned int channelIndex;
  unsigned int numberOfChannels = DSI_Headset_GetNumberOfChannels( h );
  int samplingRate = DSI_Headset_GetSamplingRate( h );

  lsl_streaminfo info;         /* out stream declaration object */
	lsl_xml_ptr desc, chn, chns; /* some xml element pointers */
	int imax = 16;
  char source_id[imax];
  getRandomString(source_id, imax);

	/* declare a new streaminfo (name: WearableSensing, content type: EEG, number of channels, srate, float values, source id*/
	info = lsl_create_streaminfo((char*)streamName,"EEG",numberOfChannels,samplingRate,cft_float32,source_id);

	/* add some meta-data fields to it */
	/* (for more standard fields, see https://github.com/sccn/xdf/wiki/Meta-Data) */
	desc = lsl_get_desc(info);
	lsl_append_child_value(desc,"manufacturer","WearableSensing");
	chns = lsl_append_child(desc,"channels");
	for( channelIndex=0; channelIndex < numberOfChannels ; channelIndex++)
  {
    chn = lsl_append_child(chns,"channel");
		lsl_append_child_value(chn,"label", (char*) DSI_Channel_GetString( DSI_Headset_GetChannelByIndex( h, channelIndex ) ) );
		lsl_append_child_value(chn,"unit","microvolts");
		lsl_append_child_value(chn,"type","EEG");
	}

	/* make a new outlet (chunking: default, buffering: 360 seconds) */
	return lsl_create_outlet(info,0,360);
}


int GlobalHelp( int argc, const char * argv[] )
{
  fprintf( stderr,
            "Usage: %s [ --OPTIONS... ]\n\n"
            "With the exception of --help,\n"
            "the options should be given in --NAME=VALUE format.\n"
            "\n"
            "  --help\n"
            "       Displays this help text.\n"
            "\n"
            "  --port\n"
            "       Specifies the serial port address (e.g. --port=COM4 on Windows,\n"
            "       --port=/dev/cu.DSI24-023-BluetoothSeri on OSX, or --port=/dev/rfcomm0 on Linux) on which to connect.\n"
            "       Note: if you omit this option, or use an empty string or the string\n"
            "       \"default\", then the API will look for an environment variable called\n"
            "       DSISerialPort and use the content of that, if available.\n"
            "\n"
            "  --montage\n"
            "       A list of channel specifications, comma-separated without spaces,\n"
            "       (can also be space-delimited, but then you would need to enclose the\n"
            "       option in quotes on the command-line).\n"
            "\n"
            "  --reference\n"
            "       The name of sensor (or linear combination of sensors, without spaces)\n"
            "       to be used as reference. Defaults to a \"traditional\" averaged-ears or\n"
            "       averaged-mastoids reference if available, or the factory reference\n"
            "       (typically Pz) if these sensors are not available.\n"
            "\n"
            "  --verbosity\n"
            "       The higher the number, the more messages the headset will send to the\n"
            "       registered `DSI_MessageCallback` function, and hence to the console\n"
            "       (and the more low-level they will tend to be)\n"
            "\n"
            "  --lsl-stream-name\n"
            "       The namem of the LSL outlet that will be created to stream the samples\n"
            "       received from the device. If omitted, the stream will be given the name WS-default.\n"
            "\n"
        , argv[ 0 ] );
        return 0;
    }


// NOTHING TO SEE HERE; MOVE ALONG.

// Helper function for figuring out command-line input flags like --port=COM4
// or /port:COM4 (either syntax will work).  Returns NULL if the specified
// option is absent. Returns a pointer to the argument value if the option
// is present (the pointer will point to '\0' if the argument value is empty
// or not supplied as part of the option string).  Better getopt
// implementations exist, but if you want something to work cross-platform
// without hassle, you have to do it yourself...

const char * GetStringOpt( int argc, const char * argv[], const char * keyword1, const char * keyword2 )
{
    int i, j;
    const char * result = NULL;
    const char * keyword;
    for( i = 1; i < argc; i++ )
    {
        int isopt = 0;
        const char * arg = argv[ i ];
        if( !arg ) continue;
        for( j = 0; arg[ j ]; j++ ) isopt |= arg[ j ] == '-' || arg[ j ] == '=' || arg[ j ] == '/' || arg[ j ] == ':';
        if( *arg == '-' || *arg == '/' ) ++arg;
        if( *arg == '-' || *arg == '/' ) ++arg;
        for( j = 0, keyword = keyword1; j < 2; j++, keyword = keyword2  )
        {
            if( keyword && strncmp( arg, keyword, strlen( keyword ) ) == 0 )
            {
                const char * potential = arg + strlen( keyword );
                if( *potential == '=' || *potential == ':' ) result = potential + 1;
                else if( *potential == '\0' || ( *keyword == '\0' && !isopt ) ) result = potential;
            }
        }
    }
    return result;
}
int GetIntegerOpt( int argc, const char * argv[], const char * keyword1, const char * keyword2, int defaultValue )
{
    char * end;
    int result;
    const char * stringValue = GetStringOpt( argc, argv, keyword1, keyword2 );
    if( !stringValue || !*stringValue ) return defaultValue;
    result = ( int ) strtol( stringValue, &end, 10 );
    if( !end || !*end ) return result;
    fprintf( stderr, "WARNING: could not interpret \"%s\" as a valid integer value for the \"%s\" option - reverting to default value %s=%g\n", stringValue, keyword1, keyword1, ( double )defaultValue );
    return defaultValue;
}
